Add stubs, round 2

Jakob37 opened the pull request:
I'll open a new PR to keep things more focused. This builds on the work from @ryanjameskennedy in #70 

Updates:

* Extracted version code into methods.
* Updated version format according to instructions by @ramsainanduri . Now it produces a single yaml file concatenating all versions.
* Individual publish dirs are removed, instead versions are output into channels and merged in a final new process step.
* Some of the previous versions was not correct - now updated.
* Reverting the "/${OUTDIR}" first slashes, as this prevents using relative paths when executing the pipeline
* Added some stubs to processed recently added to the workflow

I have tested all profiles as stub runs.

I will do some testing:

- [x] Run a wgs sample
- [x] Verify that `'''` code blocks are handled OK
- [x] Run an oncov1 sample
- [x] Update `logFile = file("/fs1/results/cron/logs/" + base + ".complete")` to use the result dir

Issues found during @alkc testing:

- [x] Look over granularity in publish dirs such that version files aren't also written there
- [x] Put "group" in the version file name (as it will be published in a shared folder)
- [x] Quote stub files
- [x] Discuss whether tab + space is OK in version code or too ugly

# Testrun of alkc's proposed PR-template

<!--
Add a description of changes below and a description of the expected outcome.
-->

What is changed? What is the result of the changes?

<!--
    Are there any issue or issues that are linked to this PR? 
    Use Github's syntax for closing keywords: `closes #1` or `fixes #1`
-->

fixes # (issue)

## Type of change

<!--
    Major change counts as a change that breaks backward compatilbity
    Minor change is a substantial change that requires testing before deployment
    Patch is a minor change like a bug fix, code comment/style fix, etc.
-->

- [x] Minor change

# Checklist:

<!--
    The checklist below applies to all types of changes. 
    Feel free to add your own items to the checklist.
-->

- [x] My code follows the [style guidelines for NF pipelines at CMD](http://mtlucmds1.lund.skane.se/wiki/doku.php?id=nextflow&s[]=nextflow#code_style_at_cmd)
- [x] I have performed a self-review of my code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] My changes generate no new warnings (Keep an eye on `.nextflow.log` !)
- [x] I have updated the CHANGELOG

<!--
    Select applicable checklist based on selection under # Type of change
    and delete the sections that do not apply to this PR:
-->

## Minor change

- [x] Stub run completes without errors or new warnings
- [x] `onco` run finishes without any new warnings/errors and the results can be loaded into scout
- [x] `wgs` single run finishes without any new warnings/errors  and the results can be loaded into scout
- [x] `wgs` trio run finishes without any new warnings/errors  and the results can be loaded into scout
- [x] At least one other person has reviewed and approved my code
- [x] I have made corresponding changes to the documentation
- [ ] Make sure last and new versions are tagged appropriately

Jakob37 commented on main.nf:
```diff
@@ -1,7 +1,8 @@
 #!/usr/bin/env nextflow
 
 // GENERAL PATHS //
-OUTDIR = params.outdir+'/'+params.subdir
```
Note to self: Restore this

Jakob37 commented on main.nf:
```diff
@@ -52,7 +53,7 @@ workflow.onComplete {
 
 	base = csv.getBaseName()
-	logFile = file("/fs1/results/cron/logs/" + base + ".complete")
+	logFile = file("${CRONDIR}/outlog.complete")
```
Note to self: Restore this (although this would be great to have in the config instead)

ramsainanduri replied:
Yes, it's better if it comes from the config, I have done similar thing in DSL2 SomaticPanelPipeline

Jakob37 replied:
Yes, at the moment it needs to be updated in each new branch to prevent emails being sent to Viktor. But I think it will be in a different PR, as it would need to be looked over how it is used in production.

Jakob37 commented on main.nf:
```diff
@@ -2541,9 +3755,93 @@ process create_yaml {
+	publishDir "${OUTDIR}/versions", mode: 'copy', overwrite: 'true', pattern: 'all_versions.yml'
+
+	input:
+		set group, versions from ch_fastp_versions.first().mix(
```
Feedback on this approach would be appreciated. The "first" is there to prevent multiple identical versions being used when the process are ran in multiple instances, i.e. just pick one version blob from each.

ramsainanduri replied:
'first' is a great function to avoid multiple identical files and is indeed a smart way to handle concurrent instances

Jakob37 replied:
OK, good to hear, thanks for the input!

alkc commented on main.nf:
```diff
@@ -788,24 +1115,43 @@ process gvcf_combine {
 	tag "$group"
 	time '20m'
-	publishDir "/${OUTDIR}/ped", mode: 'copy' , overwrite: 'true'	
+	publishDir "${OUTDIR}/ped", mode: 'copy' , overwrite: 'true'	
```
Good catch. Were the files from this process just silently not published at all or did it somehow still work?

Jakob37 replied:
It somehow worked, but you were obliged to use absolute paths (i.e. `/my/dir` would become `//my/dir` which also works).

alkc commented:
I've looked over the code and found no major issues. 

Good job. Stub runs are a great addition to the pipeline!

Jakob37 commented:
> I've looked over the code and found no major issues.
> 
> Good job. Stub runs are a great addition to the pipeline!

Great, thanks!

Did you / are you also planning to test-run @alkc ?

alkc commented:
> > I've looked over the code and found no major issues.
> > Good job. Stub runs are a great addition to the pipeline!
> 
> Great, thanks!
> 
> Did you / are you also planning to test-run @alkc ?

I'll try to at least set up a new instance and do a stub run.

alkc commented:
Humm. I tried to do a stub run now on hopper, but it fails at `create_ped`:

```
Error executing process > 'create_ped (23MD10824)'

Caused by:
  Process `create_ped (23MD10824)` terminated with an error exit status (127)

Command executed:

  touch 23MD10824_base.ped
  touch 23MD10824_ma.ped
  touch 23MD10824_fa.ped

        echo fa ma > type.val

Command exit status:
  127

Command output:
  (empty)

Command error:
  env: singularity: No such file or directory

Work dir:
  /fs1/alkc/nextflow_wgs/work/2c/0194d68cb3d9f7165f2188832541d6

Tip: you can try to figure out what's wrong by changing to the process work dir and showing the script file named `.command.sh`
```

The pipeline is started on RS-FS1 with the command:

```
module load java
module load nextflow/21.10.6
nextflow run -profile wgs -stub-run main.nf --outdir $(realpath ../results/) --crondir $(realpath ../cron/)  --csv 23MD10824_wgs-hg38.csv
```
cd /fs1/alkc/nextflow_wgs
```

alkc commented:
d'oh. forgot to load the singularity module. lemme try again.

alkc commented:
The stub run above now finishes with:

```
Completed at: 08-Nov-2023 09:33:14
Duration    : 6m 11s
CPU hours   : 8.9 (0% failed)
Succeeded   : 140
Ignored     : 1
Failed      : 1
```

The failed process is `reviewer` which fails with the following error:

```
/fs1/alkc/nextflow_wgs/work/b8/1031134a2512b1e32aeb230f603961/.command.sh: line 4: reviewer:: command not found 
```

alkc commented on main.nf:
```diff
@@ -574,17 +797,32 @@ process reviewer {
+		version_str = reviewer_version(task)
+		"""
+		touch ${id}.svg
+		echo ${version_str} > "${task.process}_versions.yml"
```
```suggestion
		echo "${version_str}" > "${task.process}_versions.yml"
```

Quoting the `version_str` var in the stub fixes the issue in my last comment.

Jakob37 replied:
OK, interesting, great that you spotted it

Jakob37 replied:
I have heard a saying that "you should never use bash variables unquoted". Seems I was bitten by that here

Jakob37 commented:
I have made the suggested update now @alkc , and testran the stubs. The reviewer one is looking OK for me now. Do you want to test further, or are you ready to approve?

I'll do a final full run of the pipeline and try loading into Scout.

alkc commented on main.nf:
```diff
@@ -845,40 +1199,63 @@ process madeline {
+	cat <<-END_VERSIONS > ${task.process}_versions.yml
+	${task.process}:
+	    ped-parser: \$(echo \$(ped_parser --version 2>&1) | sed -e "s/^.*ped_parser version: //")
+	    madeline2: \$(echo \$(madeline --version 2>&1) | grep : | sed -e"s/^.*Madeline //; s/PDE : 1.*//")
```
```suggestion
	    madeline2: \$(echo \$(madeline2 --version 2>&1) | grep : | sed -e"s/^.*Madeline //; s/PDE : 1.*//")
```

Wrong command.  Prior to this change the code above would silently fail and just print an error message for madeline in versions_all.yml instead of a version number.

Jakob37 replied:
OK, good that you spotted it. This is my mistake. Looks like I updated the key to "madeleine2" rather than the command

Jakob37 replied:
Updated

alkc commented:
Some of the individual versions.yml files get published together with results instaed of just the final versions file:

Output from one run:

```
(alkc-base) alkc@MTLUCMDS1:/fs1/alkc/results/wgs$ find . -name "*versions.yml"                                                                                                                                                                                                          ./ped/peddy_versions.yml                                                                                                                                                                                                                                                                ./ped/madeline_versions.yml                                                                                                                                                                                                                                                             ./sv_vcf/merged/svdb_merge_versions.yml                                                                                                                                                                                                                                                 ./sv_vcf/manta_versions.yml                                                                                                                                                                                                                                                             ./sv_vcf/tiddit_versions.yml                                                                                                                                                                                                                                                            ./sv_vcf/postprocessgatk_versions.yml                                                                                                                                                                                                                                                   ./plots/mito/run_haplogrep_versions.yml                                                                                                                                                                                                                                                 ./plots/mito/run_eklipse_versions.yml                                                                                                                                                                                                                                                   ./cov/chanjo_sambamba_versions.yml                                                                                                                                                                                                                                                      ./cov/gatkcov_versions.yml                                                                                                                                                                                                                                                              ./vcf/run_mutect2_versions.yml                                                                                                                                                                                                                                                          ./vcf/vcfbreakmulti_expansionhunter_versions.yml                                                                                                                                                                                                                                        ./vcf/split_normalize_versions.yml                                                                                                                                                                                                                                                      ./vcf/fastgnomad_versions.yml                                                                                                                                                                                                                                                           ./annotsv/annotsv_versions.yml                                                                                                                                                                                                                                                          ./versions/23MD10824.versions.yml                                                                                                                                                                                                                                                       
```

Jakob37 commented on main.nf:
Out of curiousity, what is the "group" name you are using? Wondering if you have a space in it, and that is why this is breaking for you, while having been fine for me.

Not saying it shouldn't be corrected, more to understand why I didn't see it before during my testing.

Jakob37 replied:
I double checked, this stub worked for me, so I am a bit confused here. But I'll add quotes (probably good to do that throughout).

Jakob37 commented on main.nf:
The intent is to tab for visual indentation and space for the indentation that will end up in the out file. What do you think - OK or too ugly? (Otherwise I guess we could think about some other way to represent it).

Jakob37 replied:
Also, the reason to move up the `${task.process}` is that otherwise it introduced a blank line in the final versions file.

This one looks a bit different from the others due to also being used within the shell using the `'''`. 

The result in my stub run is OK. Did you see some strange output here?

Jakob37 replied:
Added a comment to these ones for clarification

Jakob37 commented on main.nf:
```diff
@@ -2541,9 +3771,91 @@ process create_yaml {
+			ch_genmodscore_versions.first(),
+			ch_vcf_completion_versions.first(),
+			ch_peddy_versions.first(),
+			ch_fastgnomad_versions.first(),
+			ch_upd_versions.first(),
+			ch_roh_versions.first(),
+			ch_gatkcov_versions.first(),
+			ch_gatk_coverage_versions.first(),
+			ch_gatk_call_ploidy_versions.first(),
+			ch_gatk_call_cnv_versions.first(),
+			ch_postprocessgatk_versions.first(),
+			ch_manta_versions.first(),
+			ch_manta_panel_versions.first(),
+			ch_delly_panel_versions.first(),
+			ch_cnvkit_panel_versions.first(),
+			ch_svdb_merge_panel_versions.first(),
+			ch_tiddit_versions.first(),
+			ch_svdb_merge_versions.first(),
+			ch_annotsv_versions.first(),
+			ch_vep_sv_versions.first(),
+			ch_postprocess_vep_versions.first(),
+			ch_artefact_versions.first(),
+ 			ch_score_sv_versions.first(),
+			ch_compound_finder_versions.first()
+		).groupTuple()
+	
+	output:
+		file("all_versions.yml")
+	
+	script:
+		versions_joined = versions.join( ' ' )
+		"""
+		cat $versions_joined > all_versions.yml
+		"""
+	
+	stub:
+		versions_joined = versions.join( ' ' )
+		"""
+		cat $versions_joined > all_versions.yml
+		"""
```
Yes, that is also a good point

Jakob37 replied:
I have performed my runs locally, didn't know about the versions publishing folder. Would be great to discuss this IRL at some point.

Jakob37 commented:
> Some of the individual versions.yml files get published together with regular output files in various results dirs, instead of only being aggregated in the final versions file.
> 
> Output from one run:
> 
> ```
> (alkc-base) alkc@MTLUCMDS1:/fs1/alkc/results/wgs$ find . -name "*versions.yml"
> ./ped/peddy_versions.yml
> ./ped/madeline_versions.yml
> ./sv_vcf/merged/svdb_merge_versions.yml
> ./sv_vcf/manta_versions.yml
> ./sv_vcf/tiddit_versions.yml
> ./sv_vcf/postprocessgatk_versions.yml
> ./plots/mito/run_haplogrep_versions.yml
> ./plots/mito/run_eklipse_versions.yml
> ./cov/chanjo_sambamba_versions.yml
> ./cov/gatkcov_versions.yml
> ./vcf/run_mutect2_versions.yml
> ./vcf/vcfbreakmulti_expansionhunter_versions.yml
> ./vcf/split_normalize_versions.yml
> ./vcf/fastgnomad_versions.yml
> ./annotsv/annotsv_versions.yml
> ./versions/23MD10824.versions.yml
> ```
> 
> Edit: GAH.

OK I see. This seems to be due to the publishing dirs not being specific enough in saying what to publish, i.e. they also publish the temporary version files that are passed on to the final big merge. This will need some work to work through.

Thanks for a great reviewing round. I will continue working with these issues.

Jakob37 commented:
I have addressed the issues you spotted, and resolved some further issues I found, primarily around getting the publish dirs publishing the right files. 

I run through stubs, onco, wgs single and wgs trio, each completing without errors. The real samples could all be loaded into scout.

Do you want to take a final check before approving @alkc ?

It is hard to feel "confidence" that everything is in its right place here as there are so many moving parts in this PR, but I think we should proceed with wrapping this now and see if @ViktorHy has further feedbacks.

alkc commented:
On it!

## Re-review todo:

- [ ] Stub run onco
- [ ] Stub run wgs single
- [ ] Stub run wgs trio
- [ ] No versions files generated outside `/versions`
- [ ] No new warnings/erros in nextflow logs

Jakob37 commented:
Great @alkc !

I pushed a small change updating the logfile path to use the `params.resultsdir` as discussed. I needed to add a clause where it will add the log dir if not present. Testing it now with stubruns.

Jakob37 commented:
> I've reviewed the changes to the codebase since my last commit. Could not spot any issues!

Great, thanks for the help!

ryanjameskennedy commented:
I think eKLIPse should be 1.8. The software is no longer maintained and it does say v1-8 but I think that means v1.8.

Jakob37 commented:
> I think `eKLIPse` should be 1.8. The software is no longer maintained and it does say v1-8 but I think that means v1.8. Also `merge-vcfs` we could remove "-SNAPSHOT" perhaps?

Thanks for the comments! It looks like the eKLIPSe indeed is tagged `v1-8` (which probably means the same as 1.8, I'll update).

Yes, let's trim the -SNAPSHOT.

alkc commented:
Should the sample type-dependent code below also be in the associated stub block?

https://github.com/Clinical-Genomics-Lund/nextflow_wgs/blob/281ff34e20260bb80578511010516c8f8256b7d7/main.nf#L2253-L2256

Stub makes no distinction between sample types:

https://github.com/Clinical-Genomics-Lund/nextflow_wgs/blob/281ff34e20260bb80578511010516c8f8256b7d7/main.nf#L2268-L2276

alkc reviewed (approved):
I think I've wrapped up the testing for now. Looks pretty good!

Wgs single/trios and onco samples pass thru stub runs without new errors/warnings.

Version files are aggregated and published properly in `results/versions´.

Jakob37 commented:
> Should the sample type-dependent code below also be in the associated stub block?
> 
> https://github.com/Clinical-Genomics-Lund/nextflow_wgs/blob/281ff34e20260bb80578511010516c8f8256b7d7/main.nf#L2253-L2256
> 
> Stub makes no distinction between sample types:
> 
> https://github.com/Clinical-Genomics-Lund/nextflow_wgs/blob/281ff34e20260bb80578511010516c8f8256b7d7/main.nf#L2268-L2276

Makes sense, I'll include that

Jakob37 commented:
OK, I'll do a final round of stubs and real runs. If everything looks OK tomorrow, then I'll merge.

ryanjameskennedy reviewed (approved):
I've run the stubs and checked the results, apart from my comments, everything looks good to me!
